### ISSUE TITLE
Shallow.debug() when there are decorators on the component ?

### ISSUE THREAD

**Issue by Kamaraju333** [airbnb/enzyme#655]
```jsx
@observer @inject("exampleStore")
class ExampleComponent extends React.Component{
  render() {
    const{title}=this.props
      return (
        <div>
          <h5>{title}</h5>
          <input type=text value={exampleStore.example.value}/>
        </div>
      )
  }
}
```

when i do in my test case wrapper.debug() on shallow it gives me the component not the first child ,

how should use shallow in such cases .?

**Comment by blainekasten** [comment]
Couple different options. 

Export the component without the decorators while export default the decorated component. Then in your tests pull in the non decorated one. 

Or, you could try and query for your component in the tree. Though you'll lose the ability of using some of the root level apis.

**Comment by Kamaraju333** [comment]
@blainekasten  yeah i think that would be good option to separate it out , is it recommended to use shallow or can we test it using mount ?

**Comment by Kamaraju333** [comment]
@ljharb and thanks for response do you have any examples of the dive() ? The decorator is observer pattern which i am using and the injector is injecting the store in my component .I am using mobx

**Comment by ljharb** [comment]
I'm not familiar with that pattern, nor with mobx, nor really with decorators beyond their semantics (since they're not yet a stage 3 language feature, airbnb doesn't use them). In general, if you have a component that is created to wrap another component, `wrapper.dive()` will find the child, and re-shallow-render it for you.

**Comment by nschwan94** [comment]
I've used .dive() on the HoC, but get the MobX Error. `MobX injector: Store 'store' is not available! Make sure it is provided by some Provider`. I thought logically, that passing a store through the `options.context` optional argument might solve this, but no.

For each `ShallowWrapper` instance with .name() of `inject-{component-name}-with-store`, calling dive() yields the above MobX error, even when the optional context argument with a store is given. Not sure if I'm misunderstanding dive() or if it's simply incompatible in this case.

**Comment by ljharb** [comment]
I'm not familiar with how MobX works. If it uses child context, then #664 might be the issue. You can maybe work around it by providing the context argument in each dive call?

**Comment by nschwan94** [comment]
The thought just crossed my mind that `inject` apparently doesn't look at `context` directly to get the store it supposes is passed down `through React's context mechanism` by the `Provider` component of `mobx-react`, so this is why I presume passing in the optional context argument to `dive()` is not actually helping and the wrapped component is still throwing error because it doesn't have a reference to the `store`. Since MobX wrapped components are designed to ignore the `inject` passing of the store if a `store` prop is declared directly, could `dive()` be modified to take additional props as well as a context? You'll have to excuse me if I'm misunderstanding some things here, I'm new to the testing of MobX and even newer to Enzyme, but appreciate both.

**Comment by ljharb** [comment]
That would be very very weird for every other use case; I don't think that's a good idea.

**Comment by nschwan94** [comment]
Ok, I'll workaround.

**Comment by ljharb** [comment]
I think this issue can be closed; #664 is the issue for context via dive.
